src/async_fuse/memfs/cache/backend.rs

```diff
@@ -0,0 +1,491 @@
+//! The implementation of backend.
+
+use async_trait::async_trait;
+use clippy_utilities::OverflowArithmetic;
+use datenlord::config::{StorageParams, StorageS3Config};
+use futures::{stream, AsyncReadExt, AsyncWriteExt, StreamExt};
+use opendal::services::{Fs, S3};
+use opendal::{ErrorKind, Operator};
+use tracing::warn;
+
+use super::{Block, Storage};
+use crate::async_fuse::fuse::protocol::INum;
+
+/// A builder to build `BackendWrapper`.
+#[derive(Debug)]
+pub struct BackendBuilder {
+    /// The storage config
+    config: StorageParams,
+    /// The size of a block
+    block_size: usize,
+}
+
+impl BackendBuilder {
+    /// Create a backend builder.
+    #[must_use]
+    pub fn new(config: StorageParams, block_size: usize) -> Self {
+        Self { config, block_size }
+    }
+
+    /// Build the backend.
+    pub fn build(self) -> opendal::Result<BackendWrapper> {
+        let BackendBuilder { config, block_size } = self;
+
+        let operator = match config {
+            StorageParams::S3(StorageS3Config {
+                ref endpoint_url,
+                ref access_key_id,
+                ref secret_access_key,
+                ref bucket_name,
+            }) => {
+                let mut builder = S3::default();
+
+                builder
+                    .endpoint(endpoint_url)
+                    .access_key_id(access_key_id)
+                    .secret_access_key(secret_access_key)
+                    .region("auto")
+                    .bucket(bucket_name);
+
+                Operator::new(builder)?.finish()
+            }
+            StorageParams::Fs(ref root) => {
+                let root = if root.is_empty() {
+                    warn!("Argument `--storage-fs-root` is not set while `--storage-type` is set to `fs`. using `/tmp/datenlord_backend` as default.");
+                    "/tmp/datenlord_backend"
+                } else {
+                    root.as_str()
+                };
+
+                let mut builder = Fs::default();
+                builder.root(root);
+                Operator::new(builder)?.finish()
+            }
+        };
+
+        Ok(BackendWrapper {
+            operator,
+            block_size,
+        })
+    }
+}
+
+/// The backend wrapper of `openDAL` operator.
+#[derive(Debug)]
+pub struct BackendWrapper {
+    /// The inner `Operator`
+    operator: Operator,
+    /// Block size
+    block_size: usize,
+}
+
+impl BackendWrapper {
+    /// Create a new backend
+    #[must_use]
+    pub fn new(operator: Operator, block_size: usize) -> Self {
+        Self {
+            operator,
+            block_size,
+        }
+    }
+
+    /// Get file path by `ino`
+    fn get_file_path(ino: INum) -> String {
```

Comment:
These two functions actually have nothing to do with the backend; making them regular functions would suffice.